### PR TITLE
Type checker fixes

### PR DESCRIPTION
**Pull request by mkolosick**
This makes some correctness fixes and also has a few optimizations.

- Letrec's are now reconstructed in the correct order
- Loose check blocks are ignored by the type checker as they cause a massive code blowup
- A number of other optimizations

### CODE REVIEW

File: src/arr/compiler/type-check-structs.arr [hunk 1]
```diff
@@ -829,8 +829,10 @@ fun solve-helper-examples(system :: ConstraintSystem, solution :: ConstraintSolu
   cases(ConstraintSystem) system:
     | no-constraints => fold-result({system; solution}, context)
     | constraint-system(variables, constraints, refinement-constraints, field-constraints, example-types, next-system) =>
-      foldr-fold-result(lam(existential-key, shadow context, {shadow system; shadow solution}):
+      foldr-fold-result(lam(existential-key, shadow context, {shadow system; shadow solution}) block:
         {existential; _; fun-examples; checking-fun} = example-types.get-value(existential-key)
+        print("existential: " + tostring(existential) + "\n")
+        print("fun-examples: " + tostring(fun-examples) + "\n")
```
Comment (jswrenn) [inline src/arr/compiler/type-check-structs.arr]: On CPO, won't these `print` calls result in those messages being printed to the REPL?

Comment (mkolosick) [inline src/arr/compiler/type-check-structs.arr]: Yes, I thought I'd removed all of them but apparently I missed two.

Comment (mkolosick) [inline src/arr/compiler/type-check-structs.arr]: Wait, where are you finding those? They seemed to have been removed in 6f2614b

Comment (jswrenn) [inline src/arr/compiler/type-check-structs.arr]: Nope, you're right. The perils of viewing-by-commit.

**Pull request closed without merging**